policy-controller/k8s/index/src/outbound_index.rs

```diff
@@ -0,0 +1,361 @@
+        self.namespaces
+            .by_ns
+            .entry(ns.clone())
+            .or_insert_with(|| Namespace {
+                service_routes: Default::default(),
+                namespace: Arc::new(ns),
+                cluster_domain: self.namespaces.cluster_domain.clone(),
+                services: Default::default(),
+            })
+            .apply(route);
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        if let Some(ns_index) = self.namespaces.by_ns.get_mut(&namespace) {
+            ns_index.delete(name);
+        }
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<Service> for Index {
+    fn apply(&mut self, service: Service) {
+        let name = service.name_unchecked();
+        let ns = service.namespace().expect("Service must have a namespace");
+        if let Some(cluster_ip) = service
+            .spec
+            .as_ref()
+            .and_then(|spec| spec.cluster_ip.as_deref())
+            .filter(|ip| !ip.is_empty() && *ip != "None")
+        {
+            match cluster_ip.parse() {
+                Ok(addr) => {
+                    let service_ref = ServiceRef {
+                        name,
+                        namespace: ns.clone(),
+                    };
+                    self.services.insert(addr, service_ref);
+                }
+                Err(error) => {
+                    tracing::error!(%error, service=name, "invalid cluster ip");
+                }
+            }
+        }
+
+        let opaque_ports =
+            ports_annotation(service.annotations(), "config.linkerd.io/opaque-ports")
+                .unwrap_or_else(|| self.default_opaque_ports.clone());
+        let service_info = ServiceInfo { opaque_ports };
+
+        self.namespaces
+            .by_ns
+            .entry(ns.clone())
+            .or_insert_with(|| Namespace {
+                service_routes: Default::default(),
+                namespace: Arc::new(ns),
+                cluster_domain: self.namespaces.cluster_domain.clone(),
+                services: Default::default(),
+            })
+            .update_service(service.name_unchecked(), service_info);
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        if let Some(ns) = self.namespaces.by_ns.get_mut(&namespace) {
+            ns.services.remove(&name);
+        }
+        let service_ref = ServiceRef { name, namespace };
+        self.services.retain(|_, v| *v != service_ref);
+    }
+}
+
+impl Index {
+    pub fn shared(cluster_domain: String, default_opaque_ports: PortSet) -> SharedIndex {
+        Arc::new(RwLock::new(Self {
+            namespaces: NamespaceIndex {
+                by_ns: HashMap::default(),
+                cluster_domain: Arc::new(cluster_domain),
+            },
+            services: HashMap::default(),
+            default_opaque_ports,
+        }))
+    }
+
+    pub fn outbound_policy_rx(
+        &mut self,
+        namespace: &str,
+        service: &str,
```

Comment:
AFAICT, everywhere this method is called, we have these as owned `String`s, and we don't need them for other stuff in the calling context...why not pass them in as `String`s rather than taking them as references and then allocating new `String`s to hold the referenced string slice?